crates/matrix-sdk-base/src/client.rs

```diff
@@ -558,6 +564,44 @@ impl BaseClient {
         changes.account_data = account_data;
     }
 
+    /// User has left a room.
+    ///
+    /// Update the internal and cached state accordingly. Return the final Room.
+    pub async fn room_left(&self, room_id: &RoomId) -> Result<Room> {
+        let room = self.store.get_or_create_room(room_id, RoomType::Left).await;
+        if room.room_type() == RoomType::Left {
+            return Ok(room);
+        }
+
+        // needs to be updated first
+        room.set_room_type(RoomType::Left);
+        let room_info = room.clone_info();
+        let mut changes = StateChanges::default();
+        changes.add_room(room_info);
+        self.store.save_changes(&changes).await?;
+
+        Ok(self.store.get_or_create_room(room_id, RoomType::Left).await)
+    }
+
+    /// User has joined a room.
+    ///
+    /// Update the internal and cached state accordingly. Return the final Room.
+    pub async fn room_joined(&self, room_id: &RoomId) -> Result<Room> {
+        let room = self.store.get_or_create_room(room_id, RoomType::Joined).await;
+        if room.room_type() == RoomType::Joined {
+            return Ok(room);
+        }
+
+        // needs to be updated first
+        room.set_room_type(RoomType::Joined);
+        let room_info = room.clone_info();
+        let mut changes = StateChanges::default();
+        changes.add_room(room_info);
+        self.store.save_changes(&changes).await?;
```

Comment:
So this introduces a race, albeit one that is unlikely to cause problems.

Namely, we create a new empty base room and room info with the given state of the room. If we at the same time receive the sync response that contains the state of the room, we'll create a mostly populated room info.

Depending on who gets to write first we might overwrite the more complete info with the incomplete one, since there's no logic to heal the state, the state is lost forever for us.

I wonder if we need to guard against this using some locking around the `save_changes()` calls. Let's at least add a FIXME or something along those lines.

Reply:
yeah, I was wondering when that pops up ... that we don't have a transactional API locking a `get_or_create_with` pattern ... it is always something that comes up in databases all the time ... but is one that is rarely implemented ... we have that race twice in our code now...

Reply:
Where's the other place? Or do you mean the other place in this PR?

Reply:
I mean in this PR, yes.